// better-logging/dist/lib/formatMessage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatMessage = void 0;
const fast_safe_stringify_1 = __importDefault(require("fast-safe-stringify"));
const messageConstructionStrategy_1 = require("./enums/messageConstructionStrategy");
const constructFormattingContext = (logType, config, message) => {
    const typeColor = config.color.type[logType];
    const STAMP = (innerContent, innerColor = config.color.base) => config.color.base(`[${innerColor(innerContent)}]`);
    return ({
        msg: message,
        type: STAMP(logType, typeColor),
        time24: STAMP(new Date().toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', second: 'numeric', hour12: false })),
        time12: STAMP(new Date().toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', second: 'numeric', hour12: true })),
        date: STAMP(new Date().toLocaleString('en-UK', { year: 'numeric', month: 'numeric', day: 'numeric' })),
        unix: STAMP('' + new Date().valueOf()),
        STAMP: (content, color) => STAMP(`${content}`, color),
    });
};
const constructMessage = (strategy, args) => {
    const processArgument = (arg) => {
        if (typeof arg === 'object') {
            return fast_safe_stringify_1.default(arg);
        }
        return `${arg}`;
    };
    if (strategy === messageConstructionStrategy_1.MessageConstructionStrategy.NONE) {
        return ['', args];
    }
    if (strategy === messageConstructionStrategy_1.MessageConstructionStrategy.FIRST) {
        const [first, ...rest] = args;
        return [processArgument(first), rest];
    }
    if (strategy === messageConstructionStrategy_1.MessageConstructionStrategy.ALL) {
        return [args.map(processArgument).join(' '), []];
    }
    throw new Error(`Unknown MessageConstructionStrategy: ${strategy}`);
};
const formatMessage = (logType, config, args) => {
    const [rawMessage, remainingArgs] = constructMessage(config.messageConstructionStrategy, args);
    const formattingContext = constructFormattingContext(logType, config, rawMessage);
    const formattedMessage = config.format(formattingContext);
    return [formattedMessage, remainingArgs];
};
exports.formatMessage = formatMessage;
